lib/fuse_i.h

```diff
@@ -224,3 +227,32 @@ int fuse_loop_cfg_verify(struct fuse_loop_config *config);
 /* room needed in buffer to accommodate header */
 #define FUSE_BUFFER_HEADER_SIZE 0x1000
 
+/**
+ * Get the wanted capability flags, converting from old format if necessary
+ */
+static inline int convert_to_conn_want_ext(struct fuse_conn_info *conn,
+					   uint64_t want_ext_default,
+					   uint32_t want_default)
+{
+	/*
+	 * Convert want to want_ext if necessary.
+	 * For the high level interface this function might be called
+	 * twice, once from the high level interface and once from the
+	 * low level interface. Both, with different want_ext_default and
+	 * want_default values. In order to suppress a failure for the
+	 * second call, we check if the lower 32 bits of want_ext are
+	 * already set to the value of want.
+	 */
+	if (conn->want != want_default &&
+	    fuse_lower_32_bits(conn->want_ext) != conn->want) {
+		if (conn->want_ext != want_ext_default) {
+			fuse_log(FUSE_LOG_ERR,
+				 "fuse: both 'want' and 'want_ext' are set\n");
+			return -EINVAL;
+		}
+		conn->want_ext &= ~0xffffffffULL; /* zero lower 32 bits */
+		conn->want_ext |= conn->want; /* set lower 32 bits from want */
```

Comment:
Would this maybe read clearer?
```
conn->want_ext = (uint64_t)fuse_lower_32_bits(conn->want_ext) | conn->want;
```

Reply:
Ah, right, oops. The now-changed variant is definitely better. :slightly_smiling_face: